### ISSUE TITLE
Python 3.10.x Support

### ISSUE THREAD

**Issue by AndrewW-GIT** [AntonMu/TrainYourOwnYOLO#247]
Not sure if anyone is interested (or if this repo is still maintained), but i have modified the code to work with 

Python 3.10.2 (main, Feb 26 2022, 08:21:15) [GCC 11.2.0] on linux
matplotlib                          3.5.1
numpy                               1.22.2
keras                                  2.8.0
Keras-Preprocessing          1.1.2
opencv-python                  4.5.5.62
pandas                               1.4.1
Pillow                                 9.0.1
tensorboard                       2.8.0
tensorboard-data-server    0.6.1
tensorboard-plugin-wit      0.8.1
tensorflow                           2.8.0

Which is required to run on Ubuntu Jammy

I have only tested for my requirements, but for me it looks to work well

**Comment by AntonMu** [comment]
Thank you for that. Feel free to make a pull request. 

As one note - there are also instructions in the troubleshooting section (point 3) on how to run this code with any newer python version (including 3.10).

**Comment by thomas15738** [comment]
Hi  AndrewW-GIT, 
could you share your code for python 3.10 ? I try to run this project under python 3.10, tensorflow 2.8 and keras 2.8, but I can only train the model. It is not possible to make prediction. I can load the model, but in yolo.py detect_image(....) I get an error if I run  a session. I don't understand this error message.
The same model work well with python 3.8 and tensorflow 2.3.
I get this error:
Could not find variable conv2d_48/kernel. This could mean that the variable has been deleted. In TF1, it can also mean the variable is uninitialized. Debug info: container=localhost, status error message=Container localhost does not exist. (Could not find resource: localhost/conv2d_48/kernel)
         [[{{node conv2d_48/Conv2D/ReadVariableOp}}]]

Details:

Traceback (most recent call last):
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\.venv\lib\site-packages\tensorflow\python\client\session.py", line 1377, in _do_call
    return fn(*args)
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\.venv\lib\site-packages\tensorflow\python\client\session.py", line 1360, in _run_fn
    return self._call_tf_sessionrun(options, feed_dict, fetch_list,
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\.venv\lib\site-packages\tensorflow\python\client\session.py", line 1453, in _call_tf_sessionrun
    return tf_session.TF_SessionRun_wrapper(self._session, options, feed_dict,
tensorflow.python.framework.errors_impl.FailedPreconditionError: Could not find variable conv2d_48/kernel. This could mean that the variable has been deleted. In TF1, it can also mean the variable is uninitialized. Debug info: container=localhost, status error message=Container localhost does not exist. (Could not find resource: localhost/conv2d_48/kernel)
         [[{{node conv2d_48/Conv2D/ReadVariableOp}}]]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\3_Inference\Detector.py", line 248, in <module>
    prediction, image = detect_object(
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\Utils\utils.py", line 38, in detect_object
    prediction, new_image = yolo.detect_image(image)
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\2_Training\src\keras_yolo3\yolo.py", line 165, in detect_image
    out_boxes, out_scores, out_classes = self.sess.run(
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\.venv\lib\site-packages\tensorflow\python\client\session.py", line 967, in run
    result = self._run(None, fetches, feed_dict, options_ptr,
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\.venv\lib\site-packages\tensorflow\python\client\session.py", line 1190, in _run
    results = self._do_run(handle, final_targets, final_fetches,
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\.venv\lib\site-packages\tensorflow\python\client\session.py", line 1370, in _do_run
    return self._do_call(_run_fn, feeds, fetches, targets, options,
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\.venv\lib\site-packages\tensorflow\python\client\session.py", line 1396, in _do_call
    raise type(e)(node_def, op, message)  # pylint: disable=no-value-for-parameter
tensorflow.python.framework.errors_impl.FailedPreconditionError: Graph execution error:

Detected at node 'conv2d_48/Conv2D/ReadVariableOp' defined at (most recent call last):
    File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\3_Inference\Detector.py", line 197, in <module>
      yolo = YOLO(
    File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\2_Training\src\keras_yolo3\yolo.py", line 55, in __init__
      self.boxes, self.scores, self.classes = self.generate()
    File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\2_Training\src\keras_yolo3\yolo.py", line 90, in generate
      else yolo_body(
    File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\2_Training\src\keras_yolo3\yolo3\model.py", line 89, in yolo_body
      darknet = Model(inputs, darknet_body(inputs))
    File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\2_Training\src\keras_yolo3\yolo3\model.py", line 67, in darknet_body
      x = resblock_body(x, 1024, 4)
    File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\2_Training\src\keras_yolo3\yolo3\model.py", line 52, in resblock_body
      y = compose(
    File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\2_Training\src\keras_yolo3\yolo3\utils.py", line 21, in <lambda>
      return reduce(lambda f, g: lambda *a, **kw: g(f(*a, **kw)), funcs)
    File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\2_Training\src\keras_yolo3\yolo3\utils.py", line 21, in <lambda>
      return reduce(lambda f, g: lambda *a, **kw: g(f(*a, **kw)), funcs)
    File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\2_Training\src\keras_yolo3\yolo3\utils.py", line 21, in <lambda>
      return reduce(lambda f, g: lambda *a, **kw: g(f(*a, **kw)), funcs)
    File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\.venv\lib\site-packages\keras\engine\base_layer_v1.py", line 765, in __call__
      outputs = call_fn(cast_inputs, *args, **kwargs)
    File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\.venv\lib\site-packages\keras\layers\convolutional.py", line 248, in call
      outputs = self.convolution_op(inputs, self.kernel)
    File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\.venv\lib\site-packages\keras\layers\convolutional.py", line 233, in convolution_op
      return tf.nn.convolution(
Node: 'conv2d_48/Conv2D/ReadVariableOp'
Could not find variable conv2d_48/kernel. This could mean that the variable has been deleted. In TF1, it can also mean the variable is uninitialized. Debug info: container=localhost, status error message=Container localhost does not exist. (Could not find resource: localhost/conv2d_48/kernel)
         [[{{node conv2d_48/Conv2D/ReadVariableOp}}]]

Original stack trace for 'conv2d_48/Conv2D/ReadVariableOp':
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\3_Inference\Detector.py", line 197, in <module>
    yolo = YOLO(
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\2_Training\src\keras_yolo3\yolo.py", line 55, in __init__
    self.boxes, self.scores, self.classes = self.generate()
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\2_Training\src\keras_yolo3\yolo.py", line 90, in generate
    else yolo_body(
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\2_Training\src\keras_yolo3\yolo3\model.py", line 89, in yolo_body
    darknet = Model(inputs, darknet_body(inputs))
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\2_Training\src\keras_yolo3\yolo3\model.py", line 67, in darknet_body
    x = resblock_body(x, 1024, 4)
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\2_Training\src\keras_yolo3\yolo3\model.py", line 52, in resblock_body
    y = compose(
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\2_Training\src\keras_yolo3\yolo3\utils.py", line 21, in <lambda>
    return reduce(lambda f, g: lambda *a, **kw: g(f(*a, **kw)), funcs)
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\2_Training\src\keras_yolo3\yolo3\utils.py", line 21, in <lambda>
    return reduce(lambda f, g: lambda *a, **kw: g(f(*a, **kw)), funcs)
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\2_Training\src\keras_yolo3\yolo3\utils.py", line 21, in <lambda>
    return reduce(lambda f, g: lambda *a, **kw: g(f(*a, **kw)), funcs)
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\.venv\lib\site-packages\keras\engine\base_layer_v1.py", line 765, in __call__
    outputs = call_fn(cast_inputs, *args, **kwargs)
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\.venv\lib\site-packages\keras\layers\convolutional.py", line 248, in call
    outputs = self.convolution_op(inputs, self.kernel)
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\.venv\lib\site-packages\keras\layers\convolutional.py", line 233, in convolution_op
    return tf.nn.convolution(
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\.venv\lib\site-packages\tensorflow\python\util\traceback_utils.py", line 150, in error_handler
    return fn(*args, **kwargs)
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\.venv\lib\site-packages\tensorflow\python\util\dispatch.py", line 1082, in op_dispatch_handler
    return dispatch_target(*args, **kwargs)
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\.venv\lib\site-packages\tensorflow\python\ops\nn_ops.py", line 1148, in convolution_v2
    return convolution_internal(
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\.venv\lib\site-packages\tensorflow\python\ops\nn_ops.py", line 1280, in convolution_internal
    return op(
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\.venv\lib\site-packages\tensorflow\python\ops\nn_ops.py", line 2756, in _conv2d_expanded_batch
    return gen_nn_ops.conv2d(
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\.venv\lib\site-packages\tensorflow\python\ops\gen_nn_ops.py", line 968, in conv2d
    _, _, _op, _outputs = _op_def_library._apply_op_helper(
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\.venv\lib\site-packages\tensorflow\python\framework\op_def_library.py", line 508, in _apply_op_helper
    values = ops.convert_to_tensor(
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\.venv\lib\site-packages\tensorflow\python\profiler\trace.py", line 183, in wrapped
    return func(*args, **kwargs)
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\.venv\lib\site-packages\tensorflow\python\framework\ops.py", line 1695, in convert_to_tensor
    ret = conversion_func(value, dtype=dtype, name=name, as_ref=as_ref)
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\.venv\lib\site-packages\tensorflow\python\ops\resource_variable_ops.py", line 2098, in _dense_var_to_tensor
    return var._dense_var_to_tensor(dtype=dtype, name=name, as_ref=as_ref)  # pylint: disable=protected-access
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\.venv\lib\site-packages\tensorflow\python\ops\resource_variable_ops.py", line 1476, in _dense_var_to_tensor
    return self.value()
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\.venv\lib\site-packages\tensorflow\python\ops\resource_variable_ops.py", line 619, in value
    return self._read_variable_op()
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\.venv\lib\site-packages\tensorflow\python\ops\resource_variable_ops.py", line 728, in _read_variable_op
    result = read_and_set_handle()
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\.venv\lib\site-packages\tensorflow\python\ops\resource_variable_ops.py", line 718, in read_and_set_handle
    result = gen_resource_variable_ops.read_variable_op(
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\.venv\lib\site-packages\tensorflow\python\ops\gen_resource_variable_ops.py", line 492, in read_variable_op
    _, _, _op, _outputs = _op_def_library._apply_op_helper(
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\.venv\lib\site-packages\tensorflow\python\framework\op_def_library.py", line 740, in _apply_op_helper
    op = g._create_op_internal(op_type_name, inputs, dtypes=None,
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\.venv\lib\site-packages\tensorflow\python\framework\ops.py", line 3776, in _create_op_internal
    ret = Operation(
  File "c:\Projekte\KI-Yolo\TrainYourOwnYOLO\.venv\lib\site-packages\tensorflow\python\framework\ops.py", line 2175, in __init__
    self._traceback = tf_stack.extract_stack_for_node(self._c_op)

**Comment by AndrewW-GIT** [comment]
I've created a branch, but can not complete testing due to not being able to download the model file from the google drive location, as google thinks the file is a virus and refused. Then i ran out of time. There is another issues about this though.

An overview of the changes are

TrainYourOwnYOLO\2_Training\src\keras_yolo3\yolo.py

Remove as GPU support does not appear to be an option

from tensorflow.keras.utils import multi_gpu_model
#from tensorflow.keras.utils import multi_gpu_model

Change
#import tensorflow.python.keras.backend as K
to
import tensorflow.compat.v1.keras.backend as K


TrainYourOwnYOLO\3_Inference\Detector.py
Change
           for single_prediction in prediction:
                out_df = out_df.append(
                    pd.DataFrame(
                        [
                            [
                                os.path.basename(img_path.rstrip("\n")),
                                img_path.rstrip("\n"),
                            ]
                            + single_prediction
                            + [x_size, y_size]
                        ],
                        columns=[
                            "image",
                            "image_path",
                            "xmin",
                            "ymin",
                            "xmax",
                            "ymax",
                            "label",
                            "confidence",
                            "x_size",
                            "y_size",
                        ],
                    )
                )

to
            for single_prediction in prediction:
                out_df = pd.concat([out_df, pd.DataFrame(
                        [
                            [
                                os.path.basename(img_path.rstrip("\n")),
                                img_path.rstrip("\n"),
                            ]
                            + single_prediction
                            + [x_size, y_size]
                        ],
                        columns=[
                            "image",
                            "image_path",
                            "xmin",
                            "ymin",
                            "xmax",
                            "ymax",
                            "label",
                            "confidence",
                            "x_size",
                            "y_size",
                        ],
                    )
                    ], axis=0)

TrainYourOwnYOLO\2_Training\src\keras_yolo3\yolo3\model.py

Chanage   
grid_shape[::-1], K.dtype(feats)
to
grid_shape[...,::-1], K.dtype(feats)

and 
Change
         / K.cast(input_shape[::-1], K.dtype(feats))
to
         / K.cast(input_shape[...,::-1], K.dtype(feats))

**Comment by thomas15738** [comment]
By now, I found your branch (fork). It works well with Python 3.8 on Windows 10.
I also had problems downloading the weights from google drive. I solved it like this: I got the weights from https://pjreddie.com/media/files/yolov3.weights and saved them in the appropriate folder. For config file do
git clone https://github.com/pjreddie/darknet

In the script Download_and_Convert_YOLO_weights.py

I have replaced
```
if not os.path.isfile(os.path.join(download_folder, weights_file)):
    ....
    call_string = f"python convert.py {cfg_file} {weights_file} {h5_file}"
    subprocess.call(call_string, shell=True, cwd=download_folder)
```
with it
```
if not os.path.isfile(os.path.join(download_folder, weights_file)):
    ....
if os.path.isfile(os.path.join(download_folder, weights_file)):
    #yolov3.cfg and yolov3.weights must already exist
    call_string = f"python convert.py {cfg_file} {weights_file} {h5_file}"
    subprocess.call(call_string, shell=True, cwd=download_folder)
```

**Comment by AndrewW-GIT** [comment]
I did the same, but did not complete testing as i modified the code due to the virus warning. Plus i ran out of time

The issues you had is due to tensor flow above 2.5.x (which is also required for Python 3.10.x)

I'm glad it worked for you

I don't believe I updated the models.py changes in my branch, this is not due to python but another bug see
https://stackoverflow.com/questions/57558476/training-a-keras-model-yields-multiple-optimizer-errors
The code in this repo is similar so i applied a similar fix the the errors went away, I had to fix it as when you go about 2.4.3 ( i think) the code just crashes out randomly

**Comment by AntonMu** [comment]
A simpler solution is to follow the suggestion in the [troubleshooting section](https://github.com/AntonMu/TrainYourOwnYOLO#troubleshooting) and create a virtual python 3.7 environment.
